### PR TITLE
Apply two-way binding for BitRating component (#383)

### LINKED ISSUE #383: Missing two way binding in rating component
The component should able to update the parameter that is passed into it.

### PR DESCRIPTION
**Pull request by SomayeEbrahimi**
This closes #383

### CODE REVIEW

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs [hunk 1]
```diff
@@ -24,13 +26,27 @@ public bool IsReadonly
             }
         }
 
+        [Parameter]
+        public int Value
+        {
+            get => ratingValue;
+            set
+            {
+                ratingValue = value;
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs:35]: please check if the value is not actually changed the method returns fast.
just like the BitCheckBox component

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs [hunk 2]
```diff
@@ -12,6 +12,8 @@ public partial class BitRating
         [Parameter] public string Icon { get; set; } = "FavoriteStarFill";
         [Parameter] public string UnselectedIcon { get; set; } = "FavoriteStar";
         [Parameter] public RatingSize Size { get; set; }
+        [Parameter] public EventCallback<int> ValueChanged { get; set; }
+        [Parameter] public EventCallback<int> OnChange { get; set; }
 
         [Parameter]
         public bool IsReadonly
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs:19]: please check if the value is not actually changed the set method returns fast.
just like the BitCheckBox component

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs [hunk 3]
```diff
@@ -24,13 +26,27 @@ public bool IsReadonly
...
+        }
+
         public string[] RatingColorClass { get; set; }
         public string[] RatingIcon { get; set; }
 
         private string colorClass { get; set; } = "bit-rating-dark-fluent";
         private int min { get; set; }
 
         private bool isReadOnly;
+        private int ratingValue;
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs:49]: move these property private fields at the top of the class.
just like BitCheckBox component

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs [hunk 4]
```diff
@@ -49,6 +65,7 @@ protected override async Task OnInitializedAsync()
             min = AllowZeroStars == true ? 0 : 1;
             Max = Max > min ? Max : min;
             DefaultRating = DefaultRating > 0 ? DefaultRating : min;
+            ratingValue = DefaultRating > 0 ? DefaultRating : ratingValue;
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs:68]: are you sure about this line?
1. I think in this line the ratingValue should be the base of comparison: ratingValue = ratingValue > 0 ? ...
2. This line is changing the internal value of a 2-way bound parameter, which is not right

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs [hunk 5]
```diff
@@ -58,9 +75,14 @@ protected override async Task OnInitializedAsync()
             await base.OnInitializedAsync();
         }
 
-        protected virtual void HandleClick(int index)
+        protected virtual async Task HandleClick(int index)
         {
             FillRating(index);
+
+            if (IsEnabled is false) return;
+            Value = index;
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs:83]: this is not the correct way to change a 2-way bound parameter.
please check the implementations in the BitCheckBox component.

**Review by msynk — CHANGES_REQUESTED** [review]

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs (hunk 4 above)
Comment (SomayeEbrahimi) [reply src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs:68]: I think , if rating component has default value, for example 3, its rating value should be 3 too on Initialized method.

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Rating/BitRatingTests.cs [hunk 6]
```diff
@@ -117,39 +109,39 @@ public void BitRatingShouldRespectCustomIcon(string icon, string unselectedIcon,
...
+
+            if (clickedIndex == 0)
             {
-                defaultRating = defaultRating > 0 ? defaultRating : 1;
+                clickedIndex = 1;
             }
 
-            var bitRating = component.FindAll("button");
+            if (clickedIndex > 0 && allowZeroStars is false)
+                bitRating[clickedIndex - 1].Click();
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/Rating/BitRatingTests.cs:134]: always use { }

**Review by msynk — CHANGES_REQUESTED** [review]
Please consider resolving the non-nullable reference type warnings in this PR's changed files.

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs (hunk 1 above)
Comment (msynk) [reply src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs:34]: this is not fixed.
its required to check if the value is actually changed:

```cs
set
{
    if (value == ratingValue) return;
```

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs (hunk 3 above)
Comment (msynk) [reply src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs:48]: this one was not fixed eiter

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs (hunk 5 above)
Comment (msynk) [reply src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs:80]: again, it's not resolved

File: src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor [hunk 7]
```diff
@@ -298,15 +298,18 @@
...
+        <span>@RatingValue stars selected</span>
+    </div>
     <h5>Disabled</h5>
-    <BitRating DefaultRating="5" IsEnabled="false" />
+    <BitRating IsEnabled="false" />
     <h5>ReadOnly</h5>
-    <BitRating DefaultRating="5" IsReadonly="true" />
+    <BitRating IsReadonly="true" />
     <h5>Custom icons</h5>
     <BitRating Icon="HeartFill" UnselectedIcon="Heart" />
```
Comment (msynk) [inline src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor:314]: Please provide a sample in which the value is changing from outside of the component.
check the BitCheckBox samples, please.

**Review by msynk — CHANGES_REQUESTED** [review]

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs [hunk 8]
```diff
@@ -6,8 +6,15 @@ namespace Bit.Client.Web.BlazorUI
 {
     public partial class BitRating
     {
+
+        private bool isReadOnly;
+        private int ratingValue;
+        private string colorClass;
+        private int min;
+
+        public string[] RatingColorClass { get; set; }
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/Rating/BitRating.razor.cs:15]: Nullable Reference Type issues are still here and not fixed in this class yet!

File: src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor [hunk 9]
```diff
@@ -299,17 +299,20 @@
...
+    </div>
     <h5>Disabled</h5>
-    <BitRating DefaultRating="5" IsEnabled="false" />
+    <BitRating IsEnabled="false" @bind-Value="RatingDefaultValue" />
     <h5>ReadOnly</h5>
-    <BitRating DefaultRating="5" IsReadonly="true" />
+    <BitRating IsReadonly="true" @bind-Value="RatingDefaultValue" />
     <h5>Custom icons</h5>
-    <BitRating Icon="HeartFill" UnselectedIcon="Heart" />
+    <BitRating Icon="HeartFill" UnselectedIcon="Heart" @bind-Value="RatingDefaultValue" />
```
Comment (msynk) [inline src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor:315]: There is not any sample in which the Value is changed from outside, like the BitCheckBox sample.
for example, there could be a button and its click handler changes the value property of a BitRating.

**Review by msynk — CHANGES_REQUESTED** [review]

**Review by msynk — APPROVED** [review]

**Pull request merged**